docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,164 @@
+
+Value types may be marked with 0 (oblivious) or 1 (not annotated) - the representations are considered equivalent.
+
+### Optimizations
+To optimize trivial cases the attribute may be omitted or replaced with an attribute that takes a single byte value rather than an array. The C#8 compiler uses the following optimizations:
+
+In `NullableContextAttribute(false)` [emit context](#NullableContextAttribute):
+1. All parts are oblivious: `NullableAttribute` omitted
+1. All parts are not annotated: `NullableAttribute(1)` (rather than `new byte[] { 1 }`)
+1. All parts are annotated: `NullableAttribute(2)` (rather than `new byte[] { 2 }`)
+
+In `NullableContextAttribute(true)` [emit context](#NullableContextAttribute):
+1. All parts are oblivious: `NullableAttribute(0)` (rather than `new byte[] { 0 }`)
+1. All parts are not annotated: `NullableAttribute` omitted
+1. All parts are annotated: `NullableAttribute(2)` (rather than `new byte[] { 2 }`)
+
+The first case, where all parts are oblivious in a `NullableContextAttribute(false)` emit context, matches the representation in legacy assemblies where there are no `NullaleAttribute` annotations.
+
+### Type parameters
+The following `NullableAttribute` cases are recognized and emitted for type parameter definitions:
+
+In `NullableContextAttribute(false)` [emit context](#NullableContextAttribute):
+1. `notnull` constraint: `NullableAttribute(1)`
+1. `class` constraint in `#nullable disable`: `NullableAttribute` omitted
+1. `class` constraint in `#nullable enable`: `NullableAttribute(1)`
+1. `class?` constraint: `NullableAttribute(2)`
+1. No constraints (see below) in `#nullable disable`: `NullableAttribute` omitted
+1. No constraints (see below) in `#nullable enable`: `NullableAttribute(2)`
+
+In `NullableContextAttribute(true)` [emit context](#NullableContextAttribute):
+1. `notnull` constraint: `NullableAttribute` omitted
+1. `class` constraint in `#nullable disable`: `NullableAttribute(0)`
+1. `class` constraint in `#nullable enable`: `NullableAttribute` omitted
+1. `class?` constraint: `NullableAttribute(2)`
+1. No constraints (see below) in `#nullable disable`: `NullableAttribute(0)`
+1. No constraints (see below) in `#nullable enable`: `NullableAttribute(2)`
+
+"No constraints" (see above) means no `notnull`, `class`, `struct`, `unmanaged`, or type constraints - the equivalent of an `object?` constraint.
+
+Other forms of `NullableAttribute` on type parameter definitions will be ignored silently.
+
+### Examples
+```C#
+class A
+{
+    [Nullable(2)] string s; // string?
+    [Nullable({ 2, 1, 2 }] Dictionary<string, object> d; // Dictionary<string!, object?>?
+}
+class B
+{
+   [Nullable({ 2, 1 }] int[] a; // int[]?
+   [Nullable(1)] int[] b;       // int[]!
+   int[] c;                     // int[]~
+}
+[NullableContext(true)]
+class C
+{
+   [Nullable({ 2, 1 }] int[] a; // int[]?
+   int[] b;                     // int[]!
+   [Nullable(0)] int[] c;       // int[]~
```

Comment:
>[Nullable(0)] int[] c;       // int[]~ [](start = 3, length = 38)

For consistency with ```C.a``` I would expect to see ```[Nullable({ 0, 1 }]``` here. Unless you are proposing to make encoding for value types dependent on flags for other constituent parts within the same type reference. Please clarify.

Reply:
Yes that is the proposal. Added a comment.

---
In reply to: [290529397](https://github.com/dotnet/roslyn/pull/36152#discussion_r290529397) [](ancestors = 290529397)